File: scripts/primes.v2.py
#!/usr/bin/python

import sys
import numpy
import time
from mpi4py import MPI

def is_prime(number):
  """
    Return true if a number is prime
  """
  i = 2
  prime = True
  while i <= (number / i):
    if number % i == 0:
      prime = False
      break
    i += 1 if i == 2 else 2
  return prime

def main():
  root_process = 0
  digits_input = 0
  primes_cont = 0
  
  comm = MPI.COMM_WORLD
  rank = comm.Get_rank()
  size = comm.Get_size()

  if rank == root_process:
    start_time = time.time()
    args = sys.argv[1:]
    if len(args) == 0:
      print 'error: must specify the number digits'
      sys.exit(1)
    digits_input = int(args[0])


  digits = numpy.array(digits_input, 'i')
  comm.Bcast([digits, MPI.INT], root=root_process)

  min = 10**(digits-1)
  max = 10**digits
  interval_size = (max - min) / size
  max_interval = (rank +  1) * interval_size + min
  if rank == size: max_interval = max

  # print 'interval_size', interval_size
  # print 'numbers', rank, numpy.arange(rank*interval_size + min, max_interval, dtype='i')

  for number in numpy.arange(rank*interval_size + min, max_interval, dtype='i'):
    if is_prime(number): primes_cont += 1

  # print 'primes_cont', primes_cont
  primes = numpy.array(primes_cont, 'i')
  result = numpy.array(0, 'i')

  comm.Reduce([primes, MPI.INT], [result, MPI.INT], op=MPI.SUM, root=root_process)

  if rank == root_process:
    end_time = time.time()
    # print 'results', result, 'Time', end_time - start_time
    print'El numero de primos de', digits, 'digitos es ', result, 'Tiempo: ', end_time - start_time
  # comm.Disconnect()

main()

File: scripts/primes.v2.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#include <mpi.h>
#include <time.h>


main(int argc, char **argv) 
{
  int my_id, root_process, num_procs, ierr, digits, sum, i, j, min, max, interval_size, max_interval, primes_cont;
  clock_t start_time, end_time;
  bool is_prime;
  MPI_Status status;

  /* Let process 0 be the root process. */
  
  root_process = 0;

  /* Now replicate this process to create parallel processes. */

  ierr = MPI_Init(&argc, &argv);

  /* Find out MY process ID, and how many processes were started. */

  ierr = MPI_Comm_rank(MPI_COMM_WORLD, &my_id);
  ierr = MPI_Comm_size(MPI_COMM_WORLD, &num_procs);

  if (my_id == root_process) 
  {
    start_time = clock();
    digits = (int) strtol(argv[1], (char **)NULL, 10);
    // printf("digits: %i, num_procs: %i\n", digits, num_procs);
  }


  /* Then...no matter which process I am:
    *
    * I engage in a broadcast so that the number of intervals is 
    * sent from the root process to the other processes, and ...
    **/
  ierr = MPI_Bcast(&digits, 1, MPI_INT, root_process, MPI_COMM_WORLD);

  /* calculate the limits range */
  min = pow(10, digits - 1);
  max = pow(10, digits);

  interval_size = (int) (max - min) / num_procs;
  max_interval = (my_id + 1) * interval_size + min;

  if(my_id == num_procs)
  {
    max_interval = max;
  }

  // printf("rank: %i, min: %i, max: %i, interval: %i\n", my_id, min, max, max_interval);

  /* then calculate the sum of the areas of the rectangles for
    * which I am responsible.  Start with the (my_id +1)th
    * interval and process every num_procs-th interval thereafter.
    **/
  primes_cont = 0;
  for (i = my_id * interval_size + min; i < max_interval + 1; i += 1) 
  { 
    j = 2;
    is_prime = true;
    while (j <= i / j)
    {
      if (i % j == 0)
      {
        is_prime = false;
        break;
      }
      j += j == 2 ? 1 : 2;
    }
    if (is_prime)
    {
      primes_cont += 1;
    }
  }
  
  // printf("proc %i result: %i\n", my_id, primes_cont);

  /* and finally, engage in a reduction in which all partial sums 
    * are combined, and the grand sum appears in variable "sum" in
    * the root process,
    **/
  ierr = MPI_Reduce(&primes_cont, &sum, 1, MPI_DOUBLE, MPI_SUM, root_process, MPI_COMM_WORLD);

  /* and, if I am the root process, print the result. */

  if (my_id == root_process)
  {
    end_time = clock();
    printf("El numero de primos de %i, digitos es %i, Tiempo: %f\n", digits, sum, (float)(end_time - start_time) / CLOCKS_PER_SEC);
  } 

  /* Close down this processes. */
  ierr = MPI_Finalize();
}
